src/Algebra/Polynomial/Morphism/Morphism.agda

```diff
@@ -0,0 +1,510 @@
+-- Polynomial Morphisms
+open import Algebra.Bundles
+  using (CommutativeRing ; CommutativeSemiring; Ring;
+         SemiringWithoutAnnihilatingZero; Semiring)
+
+module ALgebra.Polynomial.Morphism.Morphism
+  {ℓ₁ ℓ₂}
+  (R : CommutativeRing ℓ₁ ℓ₂)
+  where
+
+open import Algebra.Bundles.Raw using (RawRing)
+open import Algebra.Morphism.Bundles
+  using (RingHomomorphism ; MagmaHomomorphism ; MonoidHomomorphism)
+open import Algebra.Morphism.Structures
+  using (IsRingHomomorphism ; IsMagmaHomomorphism ; IsSemiringHomomorphism;
+         IsNearSemiringHomomorphism ; IsMonoidHomomorphism)
+open import Algebra.Polynomial.Poly.Base2 R
+  using (Poly; []≈ ; ≈[]; cons ; cast-lem ; consˡ ; consʳ)
+  renaming ( sym to P-sym ; _·_ to _·P_
+           ; _+_ to _+P_ ; _*_ to _*P_; _≈_ to _≈P_; -_ to -P_)
+open import Algebra.Polynomial.Poly.Properties2 R
+  using (cast-p≈p; IsZero⇒≈0; ≈0⇒IsZero; a·p≈a∷[]*p ; *-distrib-shift)
+  renaming  (+-identityˡ to +-P-identityˡ ; +-identityʳ to +-P-identityʳ
+            ; zeroˡ to P-zeroˡ; zeroʳ to P-zeroʳ ; *-comm to *-P-comm )
+open import Algebra.Polynomial.Base2 R
+open import Algebra.Polynomial.Properties2 R as P
+  using (+-identityˡ; +-identityʳ; +-*-commutativeRing; *-zeroˡ ; *-zeroʳ)
+
+import Algebra.Definitions.RawSemiring as Def
+import Algebra.Definitions as D 
+open import Algebra.Properties.Group
+open import Algebra.Properties.Ring
+open import Data.Nat.Base as ℕ using (ℕ; suc; pred; _⊔_)
+open import Data.Vec.Base as Vec using (Vec; []; _∷_)
+open import Data.Vec.Relation.Unary.All
+open import Data.Product.Base using (proj₁ ; proj₂)
+import Level as L
+import Relation.Binary.Morphism.Structures as S
+  using (IsRelHomomorphism)
+open import Relation.Binary.PropositionalEquality.Core using (cong)
+
+module R = CommutativeRing R
+open R
+  hiding (distrib)
+  renaming
+  ( Carrier to A
+  ; _≈_ to _≈A_
+  ; _+_ to _+A_
+  ; _*_ to _*A_
+  ; -_ to -A_
+  ; 0# to A0#
+  ; 1# to A1#)
+  
+private
+  variable
+    m n k l : ℕ
+    a b c d : A
+    p : Poly m
+    q : Poly n
+    P : Polynomial
+    Q : Polynomial
+
+module PolynomialRingMorphism
+  {ℓ₃ ℓ₄}
+  (B : CommutativeRing ℓ₃ ℓ₄)
+  (β : CommutativeRing.Carrier B) 
+
+  (f : RingHomomorphism {ℓ₁} {ℓ₂}
+       (CommutativeRing.rawRing R) (CommutativeRing.rawRing B) )
+  where
+
+  module B = CommutativeRing B 
+  open B
+    using (semiringWithoutOne ; isSemiringWithoutOne ; distrib)
+    renaming
+    (_≈_ to _≈B_
+    ; _+_ to _+B_
+    ; _*_ to _*B_
+    ; -_ to -B_
+    ; setoid to B-Set)
+    using (0#; 1#)
+
+  open import Relation.Binary.Reasoning.Setoid B-Set
+    
+  open RingHomomorphism f using (⟦_⟧)
+  open D {A = B.Carrier} (_≈B_)
+    using (_MiddleFourExchange_; _DistributesOverˡ_)
+
+  eval-p : Poly m → CommutativeRing.Carrier B
+  eval-p [] = 0#
+  eval-p (a ∷ p) =  ⟦ a ⟧ +B (β *B eval-p p )
+
+
+  eval : Polynomial → CommutativeRing.Carrier B
```

Comment:
Bad name: it took me a while to figure out that this "evaluate at 1 scaled by β" or equivalently "evaluate at β". The evaluation point should be a (local) parameter, not a module parameter.